### PR TITLE
Improve queued hero/commander level-up dialog handling

### LINKED ISSUE #7467: Crash when levelup and several windows show up in a row
Thanks that last time you fix the levelup while reward showing bug. But this time I meet similar bugs.
All related to levelup window

The first one is Sanya visit an arena which is guarded by zombies (not directly fight zombies), after battle, Sanya levels up , when arena reward window needs to pop up, it crash.

Image

The second one is Sanya stay in a town with a garrison hero, I build the 1000exp building, the garrison hero levelup, after choose skill, then crash.  (I tested only garrison hero in town it doesn't crash)

Image

The third one is Roland fight the last brown flag hero, after battle,  levels up ,shoose skill, when a "brown player lose" window needs to show up then crash.

Image

[saves.zip](https://github.com/user-attachments/files/28742771/saves.zip)

[VCMI_Client_log.txt](https://github.com/user-attachments/files/28742881/VCMI_Client_log.txt)

[myportraits-test.zip](https://github.com/user-attachments/files/28743767/myportraits-test.zip) this is an empty mod without it can't get into the saves.

I have only a translation mod and a UI mod

### LINKED ISSUE #7515: [develop] AI is not ending after defeated on attack
**Describe the bug**
After human player ends the day, the AI attacks human hero, defeated, levelup, nothing happens after (see video)

**Game logs**

[VCMI_Client_log.txt.zip](https://github.com/user-attachments/files/29219748/VCMI_Client_log.txt.zip)

**To Reproduce**
1. Human player ends the day
2. The AI attacks human hero
3. AI defeated
4. Human hero levelup

**Expected behavior**
The day ends, new day starts

**Actual behavior**
Freeze

**Screenshots**
https://github.com/user-attachments/assets/0fc0d1a7-cc52-43a3-8e8e-12870f031013

**Version**

- OS: macOS ARM
- Version: 1.8.0.4a1e0662a3d8e8d66b49d08eef9467b92d119d1c

### PR DESCRIPTION
**Pull request by MichalZr6**
This is a follow-up to: 
- #7224 
- #7437
- and my earlier #7308.

It focuses on the remaining issues around queued level-up dialogs:

  - keeps queued hero / commander level-up dialogs tied to actual server-side query lifetime
  - improves chained dialog handling without reopening the earlier flicker/crash/object-visit problems

Fixes #7467
- #7467

Fixes #7515
- #7515

### CODE REVIEW

**Comment by GeorgeK1ng** [comment]
I have tried compiled artifact and level-up flickering is back. Try to use vcmiexp 9999999 a hit enter in a row, dialog flickering again, same is for hero level-up and commander level-up too.

EDIT:
And there are probably more issues - when fighting againts opponent hero - after battle are messages in weird order, now level-up, then reward - captured artifact for example, then player kill, then commander level up.

Also crash was reproduced here - https://github.com/vcmi/vcmi/actions/runs/27192111307/job/80274578114

Yes, this PR fixes something, but brings back already fixed flickering. Also breaks messages appearence order. And bring up new freeze, which I saw after hero vs hero battle, only commander level-up appeared, then freeze - don't have save right now.

**Comment by GeorgeK1ng** [comment]
I have found another freeze with builds from this PR. When I was attacked by AI on AI turn I won that battle, only hero level up dialogu was there, commanders one was missing. Will attach save and mods preset later today to easily reproduce.

**Comment by GrafDeVan** [comment]
Also found yesterday, when cheked old saves.
изображение
AI turn, I won a battle, took secondary skill and game freezed
[FreezeAfterbattle.zip](https://github.com/user-attachments/files/29337416/FreezeAfterbattle.zip)

> [2026-Jun-25 15:30:27.153322] DEBUG [runNetwork] global - Battle 0 ended for player tan
> [2026-Jun-25 15:30:27.153322] DEBUG [runNetwork] global - Battle 0 ended for player red
> [2026-Jun-25 15:30:27.186298] DEBUG [MainGUI] global - totalRedraw requested!

**Comment by MichalZr6** [comment]
Thanks for testing, will look into this later today or tomorrow.

**Comment by MichalZr6** [comment]
I fixed the issues mentioned above and some other cases as well. Logic should work now correctly but needs more testing as there might be more edge cases of some specific actions followed with a hero/commander level up.
Also I tested (both human player and AI player) those changes against my test maps (WoG enabled):
- player with two heroes garrisoned in two Dungeon towns that gain experience and level up on map start (first turn)
- player with three Dungeon towns with 3 heroes garrisoned and 3 heroes visiting that gain experience on day 2 via battle scholar academy being built via event.

File: client/windows/GUIClasses.cpp [hunk 1]
```diff
@@ -479,60 +479,98 @@ void CSplitWindow::sliderMoved(int to)
...
+		removeChild(ok.get());
+	if(mainTitle)
+		removeChild(mainTitle.get());
+	if(levelTitle)
+		removeChild(levelTitle.get());
+	if(skillIcon)
+		removeChild(skillIcon.get());
+	if(skillValue)
+		removeChild(skillValue.get());
+
```
Comment (IvanSavenko) [inline client/windows/GUIClasses.cpp]: Unnecessary I think? Resetting element should automatically unassign it from parent

File: client/windows/CCreatureWindow.cpp [hunk 2]
```diff
@@ -864,14 +867,73 @@ CStackWindow::CStackWindow(const CCommanderInstance * commander, std::vector<ui3
...
+		removeChild(commanderTab.get());
+
+	switchButtons.clear();
+	mainSection.reset();
+	activeSpellsSection.reset();
+	commanderMainSection.reset();
+	commanderBonusesSection.reset();
+	bonusesSection.reset();
+	buttonsSection.reset();
+	commanderTab.reset();
```
Comment (IvanSavenko) [inline client/windows/CCreatureWindow.cpp]: Same here - removeChild should be unnecessary

File: client/CPlayerInterface.cpp [hunk 3]
```diff
@@ -1880,6 +1909,85 @@ void CPlayerInterface::waitForAllDialogs()
...
+		if(dialog.isLevelUpDialog())
+		{
+			dialog.state = PendingDialog::State::AwaitingQueryResolution;
+			return;
+		}
+
+		dialogs.pop_front();
+
+		if(dialog.blockingPolicy == PendingDialog::Type::Blocking || showingDialog->isBusy())
+			return;
```
Comment (IvanSavenko) [inline client/CPlayerInterface.cpp:1990]: You use `dialog` after `pop_front`. 

If you are lucky - it will refer to next dialog, or deinitialized garbage. If not - crash since `pop` may reallocate vector storage.

File: client/windows/CCreatureWindow.cpp [hunk 4]
```diff
@@ -864,14 +867,73 @@ CStackWindow::CStackWindow(const CCommanderInstance * commander, std::vector<ui3
...
+	selectionSubmitted = false;
+
+	info->stackNode = commander;
+	info->creature = commander->getCreature();
+	info->commander = commander;
+	info->creatureCount = 1;
+	info->levelupInfo = std::make_optional(UnitView::CommanderLevelInfo());
+	info->levelupInfo->skills = skills;
+	info->levelupInfo->callback = callback;
+	info->owner = dynamic_cast<const CGHeroInstance *> (commander->getArmy());
```
Comment (IvanSavenko) [inline client/windows/CCreatureWindow.cpp:883]: Perhaps extract initialization to shared method and use it both in constructor and here?
(Same question for levelup window)

**Comment by GrafDeVan** [comment]
@MichalZr6 
I ran some performance tests on my old hardware. Develop be785e6fe2b279b22b918c1f7c998b0abbdc587f takes 1 minute 55 seconds while all the AI ​​from that save are running. With your build, it takes 2 minutes 46 seconds. It's clearly slower. Can you see why?
[querry.zip](https://github.com/user-attachments/files/29796349/querry.zip)

**Comment by MichalZr6** [comment]
> @MichalZr6 
> I ran some performance tests on my old hardware. Develop be785e6fe2b279b22b918c1f7c998b0abbdc587f takes 1 minute 55 seconds while all the AI ​​from that save are running. With your build, it takes 2 minutes 46 seconds. It's clearly slower. Can you see why?

Thanks for testing. This PR's changes are hardly related to AI and performance should not be the case here. Did you run more than 1 test on each branch ? The measurements may wary a lot with many AI players included.

**Comment by GrafDeVan** [comment]
I wanted to try running it from StupidAI so that battles wouldn't affect it. But the game is currently being closed during development, so I'll try a little later.

**Comment by GrafDeVan** [comment]
In develop [b6d3bd6.exe](https://download.vcmi.eu/branch/develop/windows-x64/VCMI-branch-develop-b6d3bd6.exe) , this  save contains the following server error
изображение

It's not in your build. Maybe it took longer because he was trying to fix it? Or maybe he just did more.

Another save, first day
[start.zip](https://github.com/user-attachments/files/29876990/start.zip)

develop ~16 sec
7484 ~ 19 sec

difference minimal.

By the way, this map, despite starting fast, is extremely long in the late game. It's very tricky.

File: client/windows/GUIClasses.cpp (hunk 1 above)
Comment (MichalZr6) [reply client/windows/GUIClasses.cpp]: Yes, unnecessary most likely. Fixed.

File: client/windows/CCreatureWindow.cpp (hunk 2 above)
Comment (MichalZr6) [reply client/windows/CCreatureWindow.cpp]: Fixed

File: client/windows/CCreatureWindow.cpp (hunk 4 above)
Comment (MichalZr6) [reply client/windows/CCreatureWindow.cpp:883]: Sure, done

File: client/CPlayerInterface.cpp (hunk 3 above)
Comment (MichalZr6) [reply client/CPlayerInterface.cpp:1990]: Good catch. Fixed.

**Comment by GrafDeVan** [comment]
I found a really annoying bug in the develop: when a hero levels up after a battle, they choose secondary skills, but in that battle, they kill the player. And the game crashes.
Everything works correctly in this PR.
[Rissa.zip](https://github.com/user-attachments/files/30012502/Rissa.zip)
